resolwe_bio/signals.py

```diff
@@ -56,4 +56,21 @@ def add_post_save_handler(sender, instance, **kwargs):
             for permission in list(zip(*sample._meta.permissions))[0]:  # pylint: disable=protected-access
                 shortcuts.assign_perm(permission, sample.contributor, sample)
 
+            # XXX: This doesn't work, because signal is triggered before Data
+            #      object is added to collections.
+            # for collection in Collection.objects.filter(data=instance.pk):
+            #     sample.collections.add(collection)
+
         sample.data.add(instance)
+
+
+@receiver(pre_delete, sender=Data)
+def add_pre_delete_handler(sender, instance, **kwargs):
+    """Delete Sample when last Data object is deleted."""
```

Comment:
Well, I am still thinking if we want this or not. But in any case, we have to either delete the Sample in both cases or not delete it at all.